galaxy_api/auth/auth.py

```diff
@@ -0,0 +1,109 @@
+import base64
+import json
+
+from django.conf import settings
+from django.db import transaction
+from django.db.utils import IntegrityError
+from rest_framework.authentication import BaseAuthentication
+from rest_framework.exceptions import AuthenticationFailed
+from rest_framework.permissions import BasePermission
+
+from galaxy_api.auth.models import Tenant, User
+
+
+class RHIdentityAuthentication(BaseAuthentication):
+    """
+    Authenticates users based on RedHat identity header.
+
+    For users logging in first time creates User record and
+    Tenant record for user's account if it doesn't exist.
+    """
+
+    header = 'HTTP_X_RH_IDENTITY'
+
+    def authenticate(self, request):
+        """
+        Authenticates user.
+
+        Raises:
+            AuthenticationFailed: If invalid identity header provided.
+        """
+        if self.header not in request.META:
+            return None
+
+        header = self._decode_header(request.META[self.header])
+
+        try:
+            identity = header['identity']
+            account = identity['account_number']
+
+            user = identity['user']
+            username = user['username']
+        except KeyError:
+            raise AuthenticationFailed
+
+        email = user.get('email', '')
+        first_name = user.get('first_name', '')
+        last_name = user.get('last_name', '')
+
+        tenant = self._ensure_tenant(account)
+        user = self._ensure_user(
+            username,
+            tenant,
+            email=email,
+            first_name=first_name,
+            last_name=last_name
+        )
+
+        return user, {'rh_identity': header}
+
+    @staticmethod
+    def _ensure_tenant(account):
+        try:
+            tenant, created = Tenant.objects.get_or_create(name=account)
+        except IntegrityError:
+            tenant = Tenant.objects.get(name=account)
```

Comment:
Can you explain what this is doing, because it seems redundant. Isn't the except block duplicating the functionality of the `get_or_create`?

Reply:
You're actually right. I got confused by the koku project code.
`get_or_create` handles race condition state internally.